src/tribler/core/components/user_activity/user_activity_component.py

```diff
@@ -0,0 +1,110 @@
+        self.torrent_checker: TorrentChecker = torrent_checker_component.torrent_checker
+
+        # Hook events
+        self.session.notifier.add_observer(notifications.torrent_finished, self.on_torrent_finished)
+        self.session.notifier.add_observer(notifications.remote_query_results, self.on_query_results)
+        self.session.notifier.add_observer(notifications.local_query_results, self.on_query_results)
+        self.task_manager.register_task("Check preferable", self.check_preferable,
+                                        interval=self.session.config.user_activity.health_check_interval)
+
+    def on_query_results(self, data: dict) -> None:
+        """
+        Start tracking a query and its results.
+
+        If any of the results get downloaded, we store the query (see ``on_torrent_finished``).
+        """
+        query = data.get("query")
+        if query is None:
+            return
+
+        results = {tmd.infohash for tmd in data["results"]}
+        for infohash in results:
+            self.infohash_to_queries[infohash].append(query)
+        self.queries[query] = results | self.queries.get(query, set())
+
+        if len(self.queries) > self.max_query_history:
+            query, results = self.queries.popitem(False)
+            for infohash in results:
+                self.infohash_to_queries[infohash].remove(query)
+                if not self.infohash_to_queries[infohash]:
+                    self.infohash_to_queries.pop(infohash)
+
+
+    @staticmethod
+    async def _fut_to_task(future: Future):
+        await future
+
+    def on_torrent_finished(self, infohash: str, name: str, hidden: bool) -> None:
+        """
+        When a torrent finishes, check if we were tracking the infohash. If so, store the query and its result.
+        """
+        b_infohash = InfoHash(unhexlify(infohash))
+        queries = self.infohash_to_queries[b_infohash]
+        for query in queries:
+            losing_infohashes = self.queries[query] - {b_infohash}
+            fut = get_running_loop().run_in_executor(None,
+                                                     self.database_manager.store,
+                                                     query, b_infohash, losing_infohashes)
+            self.task_manager.register_task("Store query", functools.partial(UserActivityComponent._fut_to_task, fut))
+
+    def check_preferable(self) -> None:
+        """
+        Check a preferable torrent.
+
+        This causes a chain of events that leads to the torrent being gossiped more often in the ``ContentDiscovery``
+        community.
+        """
+        random_infohashes = self.database_manager.get_random(limit=1)  # Note: this set can be empty!
+        for infohash in random_infohashes:
+            self.task_manager.register_anonymous_task("Check preferable torrent",
+                                                      self.torrent_checker.check_torrent_health, infohash)
```

Comment:
Here's a simplified version: This version avoids a for loop for a set of just a single item and avoids an unnecessary intermediate call for `register_anonymous_task`.

If you really want queuing for `check_torrent_health`, it would be better to implement it on the `check_torrent_health` side.

```suggestion
        if not random_infohashes:
            return

        infohash = random_infohashes.pop()
        await self.torrent_checker.check_torrent_health(infohash)
```

As a side note, if you never use `get_random` with a limit different from 1 (excluding tests), then it is probably a sign that the function should not return a set but a single item. See: [YAGNI on Wikipedia](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).